### ISSUE TITLE
Test count in start-run report is always 0

### ISSUE THREAD

**Issue by cyanite** [nunit/nunit-console#441]
This is because the MasterTestRunner is never "loaded" (i.e. the Load method called) before the tests are run, which then makes CountTests return 0. Curiously, no code ever calls Load except some unit tests, which thus seem to test behavior not actually encountered when running the program.

There is also a bug in Load which makes it crash for aggregated test runs.

I don't know if there is a good reason to not always call Load. It seems to work fine for me on a large-ish .nunit test suite with 1 nunit3 test assembly and 50 nunit2.

I will create a pull request which "fixes" the bug in Load (makes it return null where it otherwise crash; no one uses the result anyway), and adds an option to nunit-console, --preload, to load the MasterTestRunner before running the tests. This fixes the bug.

**Comment by cyanite** [comment]
Well, that makes sense. However @CharliePoole 's comment:

> However, it's only in the console that we limit the number of agents and the console runner doesn't need to have a count in advance.

Is only true by default. For example, there is a teamcity addin shipped with NUnit which produces special output while running tests, and I have a similar addin whose purpose is to provide progress information to a tool that's running nunit.

For our particular test workloads we won't actually end up starting a lot of processes, but I certainly appreciate that it can be a problem in general. My PR adds an command line option.

**Comment by cyanite** [comment]
Just to be clear on the above: I need the test count in order to display progress.

**Comment by CharliePoole** [comment]
@cyanite Your point is a good one. If an extension wants a valid count then we should try to give it. A command-line option feels like a hack though.  Can we allow an extension to indicate what it needs or directly ask for it instead?

**Comment by cyanite** [comment]
@CharliePoole I agree it's a bit hack-ish. Having the extension activate the load would be ideal, but I don't see how it's possible at the moment. The extension interfaces don't have a lot of methods. The one I am using is the ITestEventListener with just one method.

Also, the call to Load (in my "hack") is all the way in the console-runner executable, which otherwise just calls Run directly. I feel a command line option, while not ideal, would still be useful. Maybe a longer, less ambiguous name.

Otherwise we perhaps need something like an optional IConfiguration-something interface that extensions can implement.

**Comment by rprouse** [comment]
@cyanite thanks for the pull request, but I agree with @CharliePoole on this one, a command line option isn't the right way to go, so I closed that PR. An `IConfiguration` type interface might be the right way to go, but that will take a bit of design, so I am going to label this issue as such.

**Comment by cyanite** [comment]
@rprouse, alright, then. I'll have to keep the changes in my local build since we need progress reports on our large suites. Also note that the PR fixes another crash issue (in Load).

**Comment by cyanite** [comment]
@rprouse Of course it would be easy to add a method/property "ShouldLoad" to ITestEventListener but that would break all existing extensions.  Alternatively, a new callback-style interface could be optionally implemented, which would have a method where the test runner would be passed. But which test runner, then? Any ITestEventListeners are added at the MasterTestRunner level, but other extension points are added in different places.

**Comment by rprouse** [comment]
Or we could just provide a way for extensions to get the test count that would ensure that they tests were loaded?

**Comment by CharliePoole** [comment]
Least intrusive approach is probably declarative. Either a new attribute or a property of Extension attribute could pass info to the engine about what an extension requires. I imagine a flags enum.

**Comment by cyanite** [comment]
@CharliePoole Yes... although the problem with that is that my extension only activated when a certain environment variable is present, so as not to mess with regular console runs. With an attribute it would activate all the time. Compare to the teamcity plugin which is normally disabled (and enabled somehow, I don't know). Things like that seem to be needed since there is no way (I know of) to enable and disable plugins on the command line, is there?

Alternatively: With a new single-purpose interface, it could decide whether or not to activate dynamically:
`interface ITestRunnerConfiguration { void ConfigureRunner(ITestRunner runner); }` or something, with whatever names. "Configure" might be a too strong word. In my case, I would just call runner.Load. Maybe "BeforeTestRun" is better. (this alternative is moot if you can indeed enable and disable extensions on the command line, in which case I also vote for the declarative approach).

**Comment by cyanite** [comment]
@CharliePoole Great. As for disabling/enabling by path, note that `nunit3-console --list-extensions` don't show any paths, which would make that a bit hard to discover. I suggest doing it by name instead.

As for the optional interface, this would solve my problems nicely. The problem with an interface is that it's annoying to add methods to it later without having to change the code in all extensions. We could also add an interface with a single method returning a configuration object with a base class. Then you'd only need to recompile the extension in most cases, although that part is unavoidable of course, for an engine extension. Option 1 suggestion:
`interface IExtensionConfiguration
{
    bool Enabled { get };
    bool NeedsValidTestCaseCount { get };
}`

Or option 2:
`interface IExtensionConfiguration
{
    ExtensionConfiguration Configuration { get };
}

public class ExtensionConfiguration
{
    public virtual bool Enabled => true;
    public bool NeedsValidTestCaseCount => false;
}`

Something like that?

**Comment by CharliePoole** [comment]
@cyanite Responding to an earlier comment I missed... the "path" to an extension has nothing to do with the file system. Extension points have a "Path" that describes them uniquely. An extension is implemented by a particular type on a particular extension point. Extension points and extensions are in a many-to-many relationship, so the command-line option would be rather long.

IMO, we need a separate issue for command-line enabling / disabling if we want to do it.

**Comment by cyanite** [comment]
@CharliePoole, for background, in my extension, I activeate/deactivate it based on the presence of an environment variable. I'd prefer a command line option. I don't think we need to be able to disable/enable at the granularity of extension points (paths); I'd be happy with just disable/enable the entire extensions as a whole. Also because extensions conceivably can hook into multiple extensions points that work together.

So a command line like `... --enable-extension=Batt.NUnit3Addin.BattNUnit3Addin` (which is the name listed by --list-extensions) would suit me.

As for the interface/enum, it looks good to me.

**Comment by ChrisMaddock** [comment]
If this remains the case in the latest version of the engine, let's remove the count attribute from start-run in v4. It's inaccurate, and there's no clear efficient solution to supply this information at this point in time. It can always be added again as a future feature, if someone comes up with a solution.

**Comment by CharliePoole** [comment]
If you eliminate this, then there's no easy way for a GUI to display a progress bar.

I think we should figure out when and why this changed. It must have been after I forked the engine for the GUI, because the counts I get work for me.

**Comment by ChrisMaddock** [comment]
Sounds good. I assumed that it had never worked so nobody could be relying on it, but sounds like that's not the case.

**Comment by CharliePoole** [comment]
The OP had a very specific situation here, involving an extension. We need to run some tests... probably best to add them to the CI in fact. I'll take a look.

**Comment by CharliePoole** [comment]
I finally looked at this. The Gui doesn't have a problem because it always calls Explore before it's even possible to run and Explore forces a call to Load. But Run is supposed to force a call to Load as well, and I'm out of time right now for looking further for the moment.

**Comment by CharliePoole** [comment]
See this https://github.com/nunit/nunit-console/pull/146#issuecomment-265882511 for further info about why it is the way it is.

Possibly, we could give a valid count so long as `--agents` is not specified or is less than the number of assemblies. @cyanite @nunit/engine-team What do you think of that idea?

**Comment by CharliePoole** [comment]
@nunit/engine-team 
As we suspected, this requires a re-design for 4.0. There are two different use cases to deal with...

1. Number of agents is not limited or the limit is greater than or equal to the number of test assemblies.
2. Number of agents is limited and the limit is less than the number of test assemblies.

In case 1, there's no problem. We can easily return a valid number of test cases.

In case 2, we can only count the tests if we either ignore the limit or load and count test cases one at a time. Call those option A and option B.

With option A we violate the number of processes allowed to be active (the `--agents` limit). We could, however, allow only a subset of those processes to execute at one time. So the limit would mean "max number of test assemblies executing" at one time rather than "max number of active agents".

Option B avoids some problems but creates others. Assume we create a single process and count the tests for each assembly using that process - loading and unloading each assembly and keeping a tally. In this case, we will never exceed the `agents` limit. However, any load-time initialization like `TestCaseSource` code will execute twice, which may be a big performance hit for some people.

Alternatives I can see for V4.0...

1. Check to see whether the max agents value is specified and if it's less than the number of assemblies. If not, just count normally. If so, then either...
    1.1 Use option A
    1.2 Use option B
    1.3 Keep returning zero
2. Invent some other way to report progress, with the possibility of percent complete going down as more tests are discovered.
    2.1 Report progress as  the number of assemblies left to run.
    2.2 Two-level progress reporting, i.e. number of assemblies and progress within each assembly
    2.3 Any of the above approaches but provided as an extension by runners that need progress.
3. Stop reporting progress entirely.

All of the above are breaking changes.

I'm marking this as blocked until we decide what to do.

**Comment by ChrisMaddock** [comment]
For v4.0, I think we just remove the attribute as the minimal change for correctness. The feature has never worked, so feels more like we're adding a new feature - and that would be a separate discussion that needs someone willing to implement it.

**Comment by CharliePoole** [comment]
You suggested that earlier. As I said it's a breaking change for me if it no longer works at all - as opposed to leaving it as it is now. Obviously, that will affect our Compatibility Project.

I agree that it has never worked _for the console runner_. But if we're talking about the engine, it works depending on choices made in the runner. In the case of the console runner, we made the choice to call Run without first calling Load, which results in a zero count. We kept the attribute because __other runners__ might want to use it, calling the engine differently.

Can we separate the issue of the console from that of the engine?  If you don't like any of the other options I listed, can we resolve this by simply documenting how things work?

Regarding "a separate discussion that needs someone willing to implement it"... For some one to pop up and implement a feature, you have to first decide it's a desired feature. IMO that's what discussing the options is about. In addition to the options I listed, there are some others that were previously discussed on this thread. Can we pick a path to follow in the future, even if it's going to be in a post-4.0 release?

**Comment by CharliePoole** [comment]
I created issue #921, for documenting `<start-run>`. I think we need to do this in any case.

**Comment by ChrisMaddock** [comment]
Sorry, I missed that it was just to do with whether load had been called or not.

In that case, does that make this a simple bugfix, that doesn't need to be in 4.0?

**Comment by CharliePoole** [comment]
I think so. We made it breaking because of what the OP asked us to do about it.

Levels of non-breaking change I can see:
1. Just document better how it works.
    1.1 For runner authors (have to call Load or Explore)
    1.2 For extension authors (whether you get a valid count depends on the runner)
2. Add `CountTestCases` as another way to indirectly call Load (I was confused by the fact that it doesn't!)

**Comment by CharliePoole** [comment]
Another quite old one... I'll first summarize the background and explain _why_ we return zero.

BACKGROUND

Since multiple agents may be running assemblies, the `MasterTestRunner` creates the StartRun event, which contains a count of all the tests that will be run. In 2016, as a fix to issue #36, I added code to set the event count to '0' if the tests had not already been loaded.

The issue explains why in great detail. Essentially, the name says it all: loading assemblies in order to get a correct count was defeating the user intention in specifying the `--agents option`. We decided that any runner that actually _needed_ a correct count could call `Load` before `Run`.

In fact, however, this issue points out that some people using the Console Runner may require a correct count. This can be because of particular extensions they are using. In addition, although not previously mentioned, there may not be a problem with loading all tests before running any in cases where the number of agents has not been restricted. I say _may not_ because this would probably change overall run times to some extent.

ACTIONS

For the beta about to be released, I'll make the following changes:

1. Make sure that all lower-level runners, including agents, are loading tests in response to  a `CountTestCases` call. (I believe this has already been done.)
2. Modify the existing code in `MasterTestRunner` to force loading if `--agents` has not been specified. We'll ahve to wait and see if this produces performance problems, but the beta is likely a good place to do that.
3. Document the `--agents` option to specify that using it will cause the `StartRun` event count to be zero.
4. If all of the above fails to resolve everyone's issues fully, then we can consider additional changes such as adding a command-line option in the next beta.

@nunit/engine-team @cyanite Let me know if this makes sense!

**Comment by CharliePoole** [comment]
@cyanite If you try this and find that the command-line option is still needed, we can consider it following the first beta release.
